ddl/column.go

```diff
@@ -427,13 +427,118 @@ func (d *ddl) onSetDefaultValue(t *meta.Meta, job *model.Job) error {
 func (d *ddl) onModifyColumn(t *meta.Meta, job *model.Job) error {
 	newCol := &model.ColumnInfo{}
 	oldColName := &model.CIStr{}
-	err := job.DecodeArgs(newCol, oldColName)
+	pos := &ast.ColumnPosition{}
+	err := job.DecodeArgs(newCol, oldColName, pos)
 	if err != nil {
 		job.State = model.JobCancelled
 		return errors.Trace(err)
 	}
 
-	return errors.Trace(d.updateColumn(t, job, newCol, oldColName))
+	if pos.Tp == ast.ColumnPositionNone {
+		return errors.Trace(d.updateColumn(t, job, newCol, oldColName))
+	}
+
+	tblInfo, err := t.GetTable(job.SchemaID, job.TableID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	col := findCol(tblInfo.Columns, newCol.Name.L)
+	if col == nil {
+		col = findCol(tblInfo.Columns, oldColName.L)
+	}
+
+	originalState := col.State
+	switch col.State {
+	case model.StatePublic:
+		if err = d.modifyColumnPosition(tblInfo, newCol, oldColName, pos); err != nil {
+			job.State = model.JobCancelled
+			return errors.Trace(err)
+		}
+		// use delete only to prevent add record based on different column order.
+		// otherwise `insert into t value (1, 2)` will add different row and index record.
+		newCol.State = model.StateDeleteOnly
```

Comment:
This is because the time of observation is not the same as this problem. 
When this statement is done，we can think that all the servers are already in the same schema. And this is guaranteed by the lease. 
But if this statement hasn't done, servers may have two schemas at the same time in this cluster.  And at this point, the result of using the previous state or the next state is acceptable. @bobotu